game/domain/body.lua

```diff
@@ -61,8 +65,19 @@ function Body:setHP(hp)
   self.damage = math.max(0, math.min(self:getMaxHP() - hp, self:getMaxHP()))
 end
 
+function Body:getDefence()
+  return self:getSpec('def') + self.def_bonus
+end
+
+function Body:getDefenceDie()
+  return self:getSpec('def_die')
+end
+
 function Body:takeDamage(amount)
-  self.damage = math.min(self:getMaxHP(), self.damage + amount)
+  local defroll = RANDOM.rollDice(self:getDefence(), self:getDefenceDie())
+  local dmg = math.max(math.min(1, amount), amount - defroll)
```

Comment:
I think you can substitute the innermost `min(...)` for `1`*, because if `amount < 1` then `amount - defroll < 1` as well (because `defroll > 0` always).

*Actually, wouldn't zero be better? Thus allowing damage to be fully blocked, possibly causing zero damage?

Reply:
I did this exactly so that when damage is zero, it stays as zero. But otherwise the minimum damage is 1. While playing I noticed rolling zero a little too often, so I put the min to 1 to try and balance that.

But we can balance this by making attacks stronger too.

Reply:
I think that since we do not have a mechanic for "evading" or "dodging", the possiblity of recuding damage to zero through common means (i.e. without immunity) will be missed.

Ideally, though, we should probably discuss and test this further, since you felt it would be better this way. You can leave it like that, and I'll open an issue to investigate this later.

Reply:
i felt like i was playing tibia tbh. i was relentlessy shooting the enemy and neither of us were taking any damage in 3 out of 4 shots. i guess the attack and life stats are kinda low too, so there's that. This is all about balancing, and well, i just hate the idea of something being completely invincible just because it's high in def. Lots of RPGs don't allow zero damage unless there is immunity involved too (FF series in general comes to mind).

In the end, given other factors, dealing just one in damage is never worth it because you gon' get killed before you can beat the monster, which is clearly too much for you. But on the other hand, when there is a mob of weaklings attacking you, a ton of 1dmg might actually be worrisome.

Reply:
Yeah, it's about balancing in the end, and I see what you mean.

I said to leave it like that, but forgot to approve the changes. Will do that, then we can do some fun sciency design experiments someday to change it if necessary.

Reply:
Oh, damn, I forgot you can still remove that `min(...)`part, even reducing damage to 1.